### PR TITLE
KAFKA-6726: Fine Grained ACL for CreateTopics (KIP-277)

### PR DESCRIPTION
**Pull request by edoardocomar**
* Handling CreateTopicsRequest now requires Create auth on Topic resource and does not require Create on Cluster

* AclCommand --producer option adjusted 

* Existing Unit and Integration tests adjusted accordingly

https://issues.apache.org/jira/browse/KAFKA-6726
https://cwiki.apache.org/confluence/display/KAFKA/KIP-277+-+Fine+Grained+ACL+for+CreateTopics+API

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@gmail.com>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
Thanks for the PR, let's add an upgrade note, please.

**Comment by edoardocomar** [comment]
Hi @ijuma the PR needs to be updated to match the KIP amends

**Comment by edoardocomar** [comment]
Hi @ijuma thanks for reviewing
we rebased, marginally improved testing for auto-create and added upgrade notes done

**Comment by edoardocomar** [comment]
@ijuma last commit was just to resolve merge conflicts

**Comment by edoardocomar** [comment]
test failure unrelated (dynamic broker config)

@ijuma can you please review as the deadline for feature freeze is tomorrow

**Comment by ijuma** [comment]
This is a small feature, so there's an extra week.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 1]
```diff
@@ -815,8 +848,9 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
         assertEquals(Collections.singleton(newTopic), e.unauthorizedTopics())
     }
 
-    addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), newTopicResource)
-    addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Create)), Resource.ClusterResource)
+    // using topic-level create Acl in this test, leaving cluster-level create Acl in the matching `write` test
+    addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write), new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Create)),
+      newTopicResource)
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: It's safer to have asserts for both cases here and in the other place.

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala [hunk 2]
```diff
@@ -223,16 +235,45 @@ abstract class EndToEndAuthorizationTest extends IntegrationTestHarness with Sas
...
+  def testProduceConsumeTopicAutoCreateTopicCreateAcl(): Unit = {
+    // topic2 is not created on setup()
+    val tp2 = new TopicPartition("topic2", 0)
+    setAclsAndProduce(tp2)
+    consumers.head.assign(List(tp2).asJava)
+    consumeRecords(this.consumers.head, numRecords, topic = tp2.topic)
+  }
+
+  @Test
+  def testProduceConsumeTopicAutoCreateClusterCreateAcl(): Unit = {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala]: Do we need new tests here given `AuthorizerIntegrationTest`.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 3]
```diff
@@ -491,6 +493,37 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
...
 
+  @Test
+  def testCreateTopicAuthorizationWithClusterCreate() {
+    val requestKeyToRequest = mutable.LinkedHashMap[ApiKeys, AbstractRequest](
+      ApiKeys.CREATE_TOPICS -> createTopicsRequest
+    )
+
+    val requestKeysToAcls = Map[ApiKeys, Map[Resource, Set[Acl]]](
+      ApiKeys.CREATE_TOPICS -> clusterCreateAcl,
+    )
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: It's a bit odd that we are using maps here even though there's a single entry in each. Shall we simplify?

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, a few minor comments after the first pass.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 1 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Ok, added test cases for all code paths

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 2 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala]: yes, we think so because EndToEndAuthorizationTest uses the acl tool command line

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 3 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: agreed!

**Comment by edoardocomar** [comment]
Hi @ijuma thanks for the review - we rebased to address the conflicts and also addressed at your comments 

you will notice a weird double call to `consumers.head.poll(Duration.ofMillis(50L));`
as we wanted the test to check topic auto-creation without a write request but for some reason (unrelated to our changes) the first call to `poll` after a previously failed one did not trigger the topic auto-creation.

**Comment by edoardocomar** [comment]
about the need for a different poll timeout - I opened another jira, does not seem a critical bug 
https://issues.apache.org/jira/browse/KAFKA-6994

File: core/src/main/scala/kafka/admin/AclCommand.scala [hunk 4]
```diff
@@ -153,13 +153,16 @@ object AclCommand extends Logging {
...
+    val topicAcls = getAcl(opts, Set(Write, Describe, Create))
+    val transactionalIdAcls = getAcl(opts, Set(Write, Describe))
 
-    //Write, Describe permission on topics, Create permission on cluster, Write, Describe on transactionalIds
-    topics.map(_ -> acls).toMap[Resource, Set[Acl]] ++
-      transactionalIds.map(_ -> acls).toMap[Resource, Set[Acl]] +
-      (Resource.ClusterResource -> (getAcl(opts, Set(Create)) ++
-        (if (enableIdempotence) getAcl(opts, Set(IdempotentWrite)) else Set.empty[Acl])))
+    //Write, Describe, Create permission on topics, Write, Describe on transactionalIds
+    topics.map(_ -> topicAcls).toMap[Resource, Set[Acl]] ++
```
Comment (ijuma) [inline core/src/main/scala/kafka/admin/AclCommand.scala]: Can `toMap[Resource, Set[Acl]]` be simply `toMap`? Same for other cases.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 5]
```diff
@@ -800,27 +839,39 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
...
     val newTopicResource = new Resource(Topic, newTopic)
     addAndVerifyAcls(Set(new Acl(userPrincipal, Allow, Acl.WildCardHost, Read)), newTopicResource)
     addAndVerifyAcls(groupReadAcl(groupResource), groupResource)
-    addAndVerifyAcls(clusterAcl(Resource.ClusterResource), Resource.ClusterResource)
+    this.consumers.head.assign(List(topicPartition).asJava)
     try {
-      this.consumers.head.assign(List(topicPartition).asJava)
-      consumeRecords(this.consumers.head)
-      Assert.fail("should have thrown exception")
+      this.consumers.head.poll(Duration.ofMillis(50L));
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Why did we replace `consumeRecords` with `poll`?

**Review by ijuma — COMMENTED** [review]
A couple of quick questions while I review the rest.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 5 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: We modified the method to avoid producing any records - to avoid adding write permission which made the test more complicated than it needed to be.

So we need to pass `numRecords=0` and we had no control on the timeout, unless we made that another optional argument.
without actually consuming records it seems to me that just calling `poll` is simpler than `consumeRecords`

Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: We want to be generous with the timeout then. Any reason why it's so low?

Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: the default timeout was fine for the first call (when we expect it always to throw)

it's the second one that caused issues - see https://issues.apache.org/jira/browse/KAFKA-6994 - but that IMHO is unrelated to the changes in this PR.

if the second timeout is low - even without adding the required acl the test passed - that's the bug 6994.

if the second timeout is large the test takes longer, that's why we kept relatively short

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 6]
```diff
@@ -492,6 +495,26 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
...
   }
 
+  @Test
+  def testCreateTopicAuthorizationWithClusterCreate() {
+    removeAllAcls()
+    val resources = Set[ResourceType](Topic)
+
+    sendRequestAndVerifyResponseError(ApiKeys.CREATE_TOPICS, createTopicsRequest, resources, isAuthorized = false)
+
+    clusterCreateAcl.get(topicResource).foreach { acls =>
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: This is a bit dangerous because if `get` returns `None`, the test will pass.

File: core/src/main/scala/kafka/admin/AclCommand.scala (hunk 4 above)
Comment (edoardocomar) [reply core/src/main/scala/kafka/admin/AclCommand.scala]: thanks, existing syntax simplified

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 2 above)
Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala]: Don't we have `AclCommandTest` for that?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 5 above)
Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Did you see my comment in KAFKA-6994?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 6 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Thanks! - the test did not actually need that block, (`get` was returning `None` !) was a sloppy remainder of previous refactoring

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 5 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: yep - replied there - it's the same with the old `poll`

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 7]
```diff
@@ -1447,14 +1453,15 @@ class KafkaApis(val requestChannel: RequestChannel,
             duplicateTopics.keySet.map((_, new ApiError(Errors.INVALID_REQUEST, errorMessage))).toMap
           } else Map.empty
 
-        val completeResults = results ++ duplicatedTopicsResults
+        val unauthorizedTopicsResults =  unauthorizedTopics.keySet.map(_ -> new ApiError(Errors.TOPIC_AUTHORIZATION_FAILED, null))
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Extra space after `=`.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 8]
```diff
@@ -551,18 +566,34 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
   }
 
   @Test
-  def testCreatePermissionNeededForWritingToNonExistentTopic() {
-    val newTopic = "newTopic"
+  def testCreatePermissionOnTopicToWriteToNonExistentTopic() {
+    testCreatePermissionNeededToWriteToNonExistentTopic("newTopic",
+      Set(new Acl(userPrincipal, Allow, Acl.WildCardHost, Write), new Acl(userPrincipal, Allow, Acl.WildCardHost, Create)),
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Why do we need `Write` here?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 5 above)
Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Even then, `poll` provides little in the way of guarantees. I suggest using `TestUtils.retry` with a generous timeout and a short timeout for `poll`. This also ensures we finish as quickly as possible.

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala [hunk 9]
```diff
@@ -160,6 +171,7 @@ abstract class EndToEndAuthorizationTest extends IntegrationTestHarness with Sas
   this.serverConfig.setProperty(KafkaConfig.OffsetsTopicPartitionsProp, "1")
   this.serverConfig.setProperty(KafkaConfig.OffsetsTopicReplicationFactorProp, "3")
   this.serverConfig.setProperty(KafkaConfig.MinInSyncReplicasProp, "3")
+  this.serverConfig.setProperty(KafkaConfig.DefaultReplicationFactorProp, "3")
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:161]: Why is this needed?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 10]
```diff
@@ -1034,19 +1035,20 @@ class KafkaApis(val requestChannel: RequestChannel,
...
-    var unauthorizedForCreateTopics = Set[String]()
+    var authorizedForDescribeNotCreateTopics = Set[String]()
 
     if (authorizedTopics.nonEmpty) {
       val nonExistingTopics = metadataCache.getNonExistingTopics(authorizedTopics)
       if (metadataRequest.allowAutoTopicCreation && config.autoCreateTopicsEnable && nonExistingTopics.nonEmpty) {
         if (!authorize(request.session, Create, Resource.ClusterResource)) {
-          authorizedTopics --= nonExistingTopics
-          unauthorizedForCreateTopics ++= nonExistingTopics
+          val unauthorizedForCreateTopics = authorizedTopics.filter(topic => !authorize(request.session, Create, new Resource(Topic, topic)))
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Nit: long line.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 11]
```diff
@@ -1034,19 +1035,20 @@ class KafkaApis(val requestChannel: RequestChannel,
     var (authorizedTopics, unauthorizedForDescribeTopics) =
       topics.partition(topic => authorize(request.session, Describe, new Resource(Topic, topic)))
 
-    var unauthorizedForCreateTopics = Set[String]()
+    var authorizedForDescribeNotCreateTopics = Set[String]()
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Why did we rename this?

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 2 above)
Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala]: Generally, this test was meant to cover some high level end to end cases. It seems OK to add an auto create case, but overkill to add one for Create Topic and another for Create Cluster. We should add a test to `AclCommandTest` for the latter.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 5 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: good point - we can use `waitUntilTrue` checking for the topic being created as a consequence of `poll`

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 8 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: thanks, going to simplify by removing the constant arguments

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 9 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:161]: because we added tests which use topic auto-create, so they need a default replication factor matching the min in sync

Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:161]: thanks anyway, we spotted a missing `@Test` :(

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 11 above)
Comment (edoardocomar) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: the new name though longer and not pretty matches the what the set of topics actually is in terms of authorizations

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 2 above)
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala]: Thanks, I see the `AclCommandTest` has already coverage of adding cluster create and topic create acls. The `AuthorizerIntegrationTest` covers all request handling logic with and without auto-create.

So we only left one end2end test case for auto create with topic create acl

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 11 above)
Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: But then later we still have things like `unauthorizedForCreateTopicMetadata`. I think we should just keep this variable as it was since there is no actual change here. I'll push a minor commit.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 12]
```diff
@@ -1034,19 +1035,21 @@ class KafkaApis(val requestChannel: RequestChannel,
...
 
     if (authorizedTopics.nonEmpty) {
       val nonExistingTopics = metadataCache.getNonExistingTopics(authorizedTopics)
       if (metadataRequest.allowAutoTopicCreation && config.autoCreateTopicsEnable && nonExistingTopics.nonEmpty) {
         if (!authorize(request.session, Create, Resource.ClusterResource)) {
-          authorizedTopics --= nonExistingTopics
-          unauthorizedForCreateTopics ++= nonExistingTopics
+          val unauthorizedForCreateTopics = authorizedTopics.filter(
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: This seems like a bug, we should only be doing this check for `nonExistingTopics`. Will push a fix.

**Review by ijuma — APPROVED** [review]
I pushed a few minor changes including a merge conflict fix. This PR LGTM now and I will merge once the tests pass. @edoardocomar, please follow up if you have concerns about any of my changes.

**Comment by omkreddy** [comment]
LGTM

**Comment by edoardocomar** [comment]
thanks @ijuma

**Comment by edoardocomar** [comment]
@ijuma we looked closely at your commit dc2727e
great that you spotted that we split the wrong collection ...
we added a minor PR to cover that case: #5155

**Comment by ijuma** [comment]
@edoardocomar you read my mind as I was going to ask if you were willing to submit a PR to fix the testing gap. :)
